app/models/types/patterns/collection.rb

```diff
@@ -50,8 +55,12 @@ def all_enabled
         patterns.select { |_, pattern| pattern.enabled? }
       end
 
-      def [](value)
-        patterns.fetch(value)
+      def [](name)
+        patterns[name]
```

Comment:
I've used fetch as rather than returning `nil` on a missing key, it raises a `KeyError` not failing silently.

Reply:
As you can see in this PR, I am using it to be able to check whether a subject is defined (in a place where I don't care whether it's _disabled_). Also now that subject is optional, there are actually legitimate use cases where I ask for a certain key, but it can be missing.

On a higher level note: I am a fan of using fetch on hashes, because often an error is preferable than receiving nil.
However, I am critical about using it in a method called `[]`, because that's pretty surprising. Now our collection quacks like a duck, but when I come to feed it some bread it bites me in the finger! (in other words: It looks like the harmless hash accessor, but it behaves like the error raising hash accessor)

Reply:
I"d say that it looks like a collection accessor, but this no Hash so I wouldn't have Hash expectations on it, as it isn't even Enumerable. =)

For all I care we could have both `[]` and `fetch`. :P

Reply:
> I"d say that it looks like a collection accessor, but this no Hash so I wouldn't have Hash expectations on it

Are you telling me this doesn't qualify for principle of least surprise? :P 

But in all seriousness, things that I could imagine:

* renaming `[]` to `fetch` and adding at least `key?` for the use case of checking presence
* replacing `[](value)` by `subject()`, which returns a pattern or `nil`
    * we make wrong calls like `patterns["potato"]` or even `patterns["subject"]` impossible, which is probably the main motivation of using `fetch`
    * at the same time we allow checking presence through the same accessor

The second one is my favorite. WDYT?

Reply:
After our IRL discussion I chose the second option from the ones above.

I didn't go for the monad approach we discussed, but if you want I can give it a try in a separate commit and see how it would look.

Reply:
I think it would simplify the outside code, left an example as a comment on the review.

Reply:
Changed accordingly. I am going to resolve this discussion here, because we can continue in the other threads :)